controller/controller_server/utils.py

```diff
@@ -8,42 +10,153 @@
+    if secret_uid:
+        return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, secret_uid, obj.id))
     return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, obj.id))
 
 
-def validate_secret(secret):
-    logger.debug("validating secrets")
-    if secret:
-        if not (config.SECRET_USERNAME_PARAMETER in secret and
-                config.SECRET_PASSWORD_PARAMETER in secret and
-                config.SECRET_ARRAY_PARAMETER in secret):
+def _is_imported_string_valid(imported_string):
+    return re.match(config.SECRET_VALIDATION_REGEX, imported_string) and imported_string == imported_string.strip()
+
+
+def _validate_secret_uid(uid):
+    if not _is_imported_string_valid(uid):
+        raise ValidationException(messages.invalid_uid_value_message.format(uid))
+    if len(uid) > config.SECRET_UID_MAX_LENGTH:
+        raise ValidationException(
+            messages.parameter_length_is_too_long.format("secret_uid", uid, config.SECRET_UID_MAX_LENGTH))
+
+
+def _validate_secret(secret):
+    if not (config.SECRET_USERNAME_PARAMETER in secret and
+            config.SECRET_PASSWORD_PARAMETER in secret and
+            config.SECRET_ARRAY_PARAMETER in secret):
+        raise ValidationException(messages.invalid_secret_message)
+
+
+def _validate_topologies(secret_topologies):
+    for topologies in secret_topologies:
+        for topology_key, topology_value in topologies.items():
+            split_topology = topology_key.split(config.PARAMETERS_TOPOLOGY_DELIMITER)
+            if len(split_topology) != 2:
+                raise ValidationException(messages.invalid_topology_key_message.format(topology_key))
+            prefix, suffix = split_topology
+            if not _is_imported_string_valid(suffix):
+                raise ValidationException(messages.invalid_topology_key_message.format(suffix))
+            if prefix not in config.TOPOLOGY_PREFIXES:
+                raise ValidationException(messages.invalid_topology_key_message.format(prefix))
+            if not _is_imported_string_valid(topology_value):
+                raise ValidationException(messages.invalid_topology_value_message.format(topology_value))
+
+
+def _validate_secret_config(secret_config_string):
+    secret_config = ast.literal_eval(secret_config_string)
+    for uid, secret_info in secret_config.items():
+        if uid and secret_info:
+            _validate_secret_uid(uid)
+            _validate_secret(secret_info)
+            topologies = secret_info.get(config.SECRET_SUPPORTED_TOPOLOGIES_PARAMETER)
+            if topologies:
+                _validate_topologies(topologies)
+        else:
             raise ValidationException(messages.invalid_secret_message)
 
-    else:
-        raise ValidationException(messages.secret_missing_message)
 
+def validate_secrets(secrets):
+    logger.debug("validating secrets")
+    if not secrets:
+        raise ValidationException(messages.secret_missing_message)
```

Comment:
then why do we usually do `if request.secrets:` before calling this function?

Reply:
I guess the `if request.secrets:` is redundant and should be removed. (as it fixed in [validate volume capabilities](https://github.com/IBM/ibm-block-csi-driver/pull/212/files#diff-cfa9702e443362434f95936cee6df91703ec484838eb64262231cab9fd03466bR156))